### PR TITLE
Swift 3

### PR DESCRIPTION
**Pull request by RishabhTayal**

### CODE REVIEW

**Comment by RishabhTayal** [comment]
Building is failing because travis doesn't support the Xcode 8 beta yet.

File: ConsistencyManager/ConsistencyManager.swift [hunk 1]
```diff
@@ -243,7 +243,7 @@ public class ConsistencyManager {
             // So, we should just return
             return
         }
-        if let newModel = newModel where outdatedModel.isEqualToModel(newModel) {
+        if let newModel = newModel , outdatedModel.isEqualToModel(newModel) {
```
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift]: small - remove space before ,

File: ConsistencyManager/ConsistencyManager.swift [hunk 2]
```diff
@@ -416,12 +416,13 @@ public class ConsistencyManager {
         if garbageCollectionInterval > 0 {
             // We're going to use a low priority queue for this timer
             // We need to use a dispatch queue because NSOperationQueue doesn't support delays
-            let dispatchQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0)
+            
+            let dispatchQueue = DispatchQueue.global(qos: DispatchQoS.QoSClass.default)
```
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift]: Looks like we've moved from low to default? Is there no low in the new API?

File: ConsistencyManager/ConsistencyManager.swift [hunk 3]
```diff
@@ -416,12 +416,13 @@ public class ConsistencyManager {
...
             // Weak here is necessary, otherwise, we'd have a retain cycle.
-            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, Int64(UInt64(garbageCollectionInterval) * NSEC_PER_SEC)), dispatchQueue) { [weak self] in
+            dispatchQueue.asyncAfter(deadline: DispatchTime.now() + Double(Int64(UInt64(garbageCollectionInterval) * NSEC_PER_SEC)) / Double(NSEC_PER_SEC)) { [weak self] in
```
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift:421]: `Double(Int64(UInt64(garbageCollectionInterval) * NSEC_PER_SEC)) / Double(NSEC_PER_SEC))`
Looks like we're multiplying by NSEC_PER_SEC then dividing again?
Does `garbageCollectionInterval` do the same thing here?

Also, DispatchTime.now() is now an `TimeInterval`?

File: ConsistencyManager/ConsistencyManager.swift [hunk 4]
```diff
@@ -610,19 +611,19 @@ public class ConsistencyManager {
...
-        dispatch_async(dispatch_get_main_queue()) {
+        DispatchQueue.main.async {
             results.forEach { (listener, newModel, modelUpdates) in
                 // Let's make sure the model hasn't changed since we last requested it.
                 // We're just going to make sure the id is the same so it represents the same data.
                 // If the id is nil, then we will still update since this represents a delete.
                 if let id = newModel?.modelIdentifier,
                     let currentId = listener.currentModel()?.modelIdentifier
-                    where id != currentId {
+                    , id != currentId {
```
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift]: I think it'd be better to put the comma on the previous line?

File: ConsistencyManager/ConsistencyManager.swift [hunk 5]
```diff
@@ -708,20 +709,20 @@ public class ConsistencyManager {
...
-    private func changedSubmodelIdsFromModel(model: ConsistencyManagerModel, modelUpdates: [String: [ConsistencyManagerModel]?]) -> Set<String> {
+    private func changedSubmodelIdsFromModel(_ model: ConsistencyManagerModel, modelUpdates: [String: [ConsistencyManagerModel]?]) -> Set<String> {
         var changedModels = Set<String>()
         model.forEach { child in
             if let id = child.modelIdentifier, let update = modelUpdates[id] {
                 // Update is still an optional because the value of model updates is optional
                 // We can ignore deletes because we are only looking for updated models.
                 // Here, we should merge and check for equality to see if anything has actually changed.
-                if let update = update where !self.mergedModelFromModel(child, withUpdates: update).isEqualToModel(child) {
+                if let update = update , !self.mergedModelFromModel(child, withUpdates: update).isEqualToModel(child) {
```
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift]: small, remove space before comma

File: ConsistencyManager/ConsistencyManager.swift (hunk 2 above)
Comment (RishabhTayal) [inline ConsistencyManager/ConsistencyManager.swift]: seems like `DISPATCH_QUEUE_PRIORITY_LOW` has become `.utility` in swift 3. I will correct that.

File: ConsistencyManager/ConsistencyManager.swift [hunk 6]
```diff
@@ -388,7 +388,7 @@ public class ConsistencyManager {
     */
     public func cleanMemory() {
         dispatchTask { _ in
-            NSNotificationCenter.defaultCenter().postNotificationName(ConsistencyManager.kCleanMemoryAsynchronousWorkStarted, object: self)
+            NotificationCenter.default.post(name: Notification.Name(rawValue: ConsistencyManager.kCleanMemoryAsynchronousWorkStarted), object: self)
```
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift]: Is this the right way to do it? Shouldn't we instead create a new Notification.Name? Guessing though...not familiar with swift 3 yet.

File: ConsistencyManager/DataStructures/BatchListener.swift [hunk 7]
```diff
@@ -75,17 +75,17 @@ public class BatchListener: ConsistencyManagerListener {
         return BatchUpdateModel(models: models)
     }
 
-    public func modelUpdated(model: ConsistencyManagerModel?, updates: ModelUpdates, context: Any?) {
-        if let model = model as? BatchUpdateModel where model.models.count == listeners.count {
+    public func modelUpdated(_ model: ConsistencyManagerModel?, updates: ModelUpdates, context: Any?) {
+        if let model = model as? BatchUpdateModel , model.models.count == listeners.count {
```
Comment (plivesey) [inline ConsistencyManager/DataStructures/BatchListener.swift]: space

File: ConsistencyManager/ConsistencyManager.swift (hunk 3 above)
Comment (RishabhTayal) [inline ConsistencyManager/ConsistencyManager.swift:421]: Not sure about the logic in this line. Please make the appropriate changes yourself.

File: ConsistencyManager/DataStructures/WeakArray.swift [hunk 8]
```diff
@@ -117,14 +117,19 @@ public struct WeakArray<T: AnyObject>: ArrayLiteralConvertible {
...
     // Required by SequenceType
-    public func generate() -> IndexingGenerator<WeakArray<T>> {
+    public func makeIterator() -> IndexingIterator<WeakArray<T>> {
         // Rather than implement our own generator, let's take advantage of the generator provided by IndexingGenerator
-        return IndexingGenerator<WeakArray<T>>(self)
+        return IndexingIterator<WeakArray<T>>(_elements: self)
     }
 
+    public func index(after i: Int) -> Int {
+        //TODO: modify the code to return the correct value
```
Comment (plivesey) [inline ConsistencyManager/DataStructures/WeakArray.swift]: Why is this necessary? And how hard is it to implement this correctly?

File: ConsistencyManager.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -522,6 +522,7 @@
 				PRODUCT_NAME = "$(TARGET_NAME)";
```
Comment (plivesey) [inline ConsistencyManager.xcodeproj/project.pbxproj:522]: I presume tests pass locally? And the sample app runs?

File: ConsistencyManager/ConsistencyManager.swift (hunk 6 above)
Comment (RishabhTayal) [inline ConsistencyManager/ConsistencyManager.swift]: This seems like the right way to do it. There could be an alternative where we initialize the `kCleanMemoryAsynchronousWorkStarted` variable of type `Notification.Name`. So the code will change to this:

```
public static let kCleanMemoryAsynchronousWorkStarted = Notification.Name("com.linkedin.consistencyManager.kCleanMemoryAsynchronousWorkStarted")
```

```
NotificationCenter.default.post(name: ConsistencyManager.kCleanMemoryAsynchronousWorkStarted, object: self)
```

Which one is preferred?

File: ConsistencyManager/ConsistencyManager.swift (hunk 3 above)
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift:421]: You added an extra `/ Double(NSEC_PER_SEC)` from the previous diff. Was this intentional? Created #28 btw, we can look at this later (don't need to block this).

File: ConsistencyManager/DataStructures/WeakArray.swift (hunk 8 above)
Comment (RishabhTayal) [inline ConsistencyManager/DataStructures/WeakArray.swift]: This is a required method for the protocol implementation. I believe this is the right implementation. I wasn't completely sure so I just added comment.

**Comment by plivesey** [comment]
Yeah, expected travis to fail. No worries.
Btw, are you planning on using this in a swift 3 project? Or just contributing to help out?

File: ConsistencyManager.xcodeproj/project.pbxproj (hunk 9 above)
Comment (RishabhTayal) [inline ConsistencyManager.xcodeproj/project.pbxproj:522]: Haven't run the tests. Sample app builds successfully.

File: ConsistencyManager/ConsistencyManager.swift (hunk 3 above)
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift:421]: Intended behavior: it should delay for `garbageCollectionInterval` seconds btw.

**Comment by RishabhTayal** [comment]
@plivesey Currently I don't see this library being used in any of my projects. I am just contributing. :)

File: ConsistencyManager/ConsistencyManager.swift (hunk 6 above)
Comment (plivesey) [inline ConsistencyManager/ConsistencyManager.swift]: I think this seems more swifty to me? So convert the static var to `Notification.Name(...`

File: ConsistencyManager/DataStructures/WeakArray.swift (hunk 8 above)
Comment (plivesey) [inline ConsistencyManager/DataStructures/WeakArray.swift]: Oh...yeah, I see now. Yeah, I think this is the right implementation too. I think it's ok to remove this TODO. If tests pass, then I think it's safe to assume this is correct.

**Comment by plivesey** [comment]
From http://stackoverflow.com/questions/37801370/how-do-i-dispatch-sync-dispatch-async-dispatch-after-etc-in-swift-3 it looks like the new dispatch_after logic is correct.

**Comment by RishabhTayal** [comment]
@plivesey made all the above mentioned changes. 😃

**Comment by plivesey** [comment]
Have you ran tests yet? Do they pass for you?

**Comment by RishabhTayal** [comment]
@plivesey I don't think tests are running properly. I am getting some build errors in test target.

**Comment by RishabhTayal** [comment]
@plivesey btw just saw you wrote an multi peer app https://github.com/plivesey/PLPartyTime. I created something like this a while ago: https://github.com/RishabhTayal/ChatApp. 😸

**Comment by plivesey** [comment]
Going to try running tests locally. See if my beta xcode is good enough...

**Comment by plivesey** [comment]
Actually, we're committing to a branch anyway. We can fix tests later.
My Xcode beta version is out of date.
Would you prefer to push this first?

**Comment by plivesey** [comment]
its just a library actually to get peer to peer without invitations. I was going to build a live game with it (similar to SpaceTeam - if you haven't played it, its amazing).

**Comment by RishabhTayal** [comment]
I think you should be able to merge this the new branch and check the tests later. 

I haven't played SpaceTeam. Will check it out.
